File: RootTools2/src/main/java/com/rarnu/tools/neo/comp/PreferenceEx.java
package com.rarnu.tools.neo.comp;

import android.content.Context;
import android.content.res.TypedArray;
import android.preference.Preference;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.Switch;
import android.widget.TextView;
import com.rarnu.tools.neo.R;

public class PreferenceEx extends Preference {

    private RelativeLayout layPref = null;
    private ImageView prefIcon = null;
    private TextView prefTitle = null;
    private TextView prefSummary = null;
    private Switch prefStatus = null;

    private View innerView = null;
    private boolean showSwitch = false;
    private boolean showIcon = true;
    private boolean isOn = false;

    public PreferenceEx(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initAttr(attrs);
    }

    public PreferenceEx(Context context, AttributeSet attrs) {
        super(context, attrs);
        initAttr(attrs);
    }

    public PreferenceEx(Context context) {
        super(context);
    }

    private void initAttr(AttributeSet attrs) {
        if (attrs != null) {
            TypedArray a = getContext().obtainStyledAttributes(attrs, R.styleable.PreferenceEx, 0, 0);
            showSwitch = a.getBoolean(R.styleable.PreferenceEx_showSwitch, false);
            showIcon = a.getBoolean(R.styleable.PreferenceEx_showIcon,true);
            a.recycle();
        }
    }

    @Override
    protected void onBindView(View view) {
        try {
            super.onBindView(view);
        } catch (Exception e) {

        }
        prefTitle.setText(getTitle());
        prefSummary.setText(getSummary());
        if (getSummary() == null || getSummary().equals("")) {
            prefSummary.setVisibility(View.GONE);
        }
        prefStatus.setChecked(isOn);
        prefIcon.setImageDrawable(getIcon());
    }

    @Override
    protected View onCreateView(ViewGroup parent) {
        super.onCreateView(parent);
        if (innerView == null) {
            innerView = LayoutInflater.from(getContext()).inflate(R.layout.comp_preference, parent, false);
            layPref = (RelativeLayout) innerView.findViewById(R.id.layPref);
            prefIcon = (ImageView) innerView.findViewById(R.id.prefIcon);
            prefTitle = (TextView) innerView.findViewById(R.id.prefTitle);
            prefSummary = (TextView) innerView.findViewById(R.id.prefSummary);
            prefStatus = (Switch) innerView.findViewById(R.id.prefStatus);
            prefStatus.setVisibility(showSwitch ? View.VISIBLE : View.GONE);
            prefIcon.setVisibility(showIcon ? View.VISIBLE : View.GONE);
        }
        return innerView;
    }

    @Override
    public void setTitle(int titleResId) {
        super.setTitle(titleResId);
        prefTitle.setText(titleResId);
    }

    @Override
    public void setSummary(int summaryResId) {
        super.setSummary(summaryResId);
        prefSummary.setText(summaryResId);
        if (getSummary() == null || getSummary().equals("")) {
            prefSummary.setVisibility(View.GONE);
        }
    }

    @Override
    public void setIcon(int iconResId) {
        super.setIcon(iconResId);
        prefIcon.setImageDrawable(getIcon());
    }

    public void setShowSwitch(boolean on) {
        prefStatus.setVisibility(on ? View.VISIBLE : View.GONE);
    }

    public void setShowIcon(boolean on) {
        prefIcon.setVisibility(on ? View.VISIBLE : View.GONE);
    }

    public void setStatus(boolean on) {
        isOn = on;
        if (prefStatus != null) {
            prefStatus.setChecked(on);
        }
    }

    public boolean getStatus() {
        return prefStatus.isChecked();
    }
}

File: RootTools2/src/main/kotlin/com/rarnu/tools/neo/comp/PreferenceEx.kt
package com.rarnu.tools.neo.comp

import android.content.Context
import android.preference.Preference
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.rarnu.tools.neo.R
import kotlinx.android.synthetic.main.comp_preference.view.*

class PreferenceEx : Preference {

    private var innerView: View? = null
    private var showSwitch = false
    private var showIcon = true
    private var exTitle: String? = ""
    private var isOn = false

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        initAttr(attrs)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        initAttr(attrs)
    }

    constructor(context: Context) : super(context) {
        initAttr(null)
    }

    private fun initAttr(attrs: AttributeSet?) {
        if (attrs != null) {
            val a = context.obtainStyledAttributes(attrs, R.styleable.PreferenceEx, 0, 0)
            showSwitch = a.getBoolean(R.styleable.PreferenceEx_showSwitch, false)
            showIcon = a.getBoolean(R.styleable.PreferenceEx_showIcon, true)
            exTitle = a.getString(R.styleable.PreferenceEx_extitle)
            a.recycle()
        }
    }

    override fun onBindView(view: View) {
        try {
            super.onBindView(view)
        } catch (e: Exception) {

        }

        innerView?.prefTitle?.text = title
        innerView?.prefExTitle?.text = exTitle
        innerView?.prefSummary?.text = summary
        if (summary == null || summary == "") {
            innerView?.prefSummary?.visibility = View.GONE
        }
        innerView?.prefStatus?.isChecked = isOn
        innerView?.prefIcon?.setImageDrawable(icon)
    }

    override fun onCreateView(parent: ViewGroup): View? {
        super.onCreateView(parent)
        if (innerView == null) {
            innerView = LayoutInflater.from(context).inflate(R.layout.comp_preference, parent, false)
            innerView?.prefStatus?.visibility = if (showSwitch) View.VISIBLE else View.GONE
            innerView?.prefIcon?.visibility = if (showIcon) View.VISIBLE else View.GONE
        }
        return innerView
    }

    override fun setTitle(titleResId: Int) {
        super.setTitle(titleResId)
        innerView?.prefTitle?.setText(titleResId)
    }

    fun setExTitle(resId: Int) {
        innerView?.prefExTitle?.setText(resId)
    }

    override fun setSummary(summaryResId: Int) {
        super.setSummary(summaryResId)
        if (summary == null || summary == "") {
            innerView?.prefSummary?.visibility = View.GONE
        }
    }

    override fun setIcon(iconResId: Int) {
        super.setIcon(iconResId)
        innerView?.prefIcon?.setImageDrawable(icon)
    }

    fun setShowSwitch(on: Boolean) {
        showSwitch = on
        innerView?.prefStatus?.visibility = if (on) View.VISIBLE else View.GONE
    }

    fun setShowIcon(on: Boolean) {
        innerView?.prefIcon?.visibility = if (on) View.VISIBLE else View.GONE
    }

    var status: Boolean
        get() = innerView?.prefStatus!!.isChecked
        set(on) {
            isOn = on
            innerView?.prefStatus?.isChecked = on
        }
}
